### PR TITLE
LowStar libraries for `vector` and the `regional` typeclass

### PR DESCRIPTION
**Pull request by joonwonc**
LowStar libraries for `vector` and the `regional` typeclass:
- LowStar.Vector.fst: `vector`, a resizable `LowStar.Buffer` where a struct has its current size and capacity
- LowStar.Regional.fst: the `regional` typeclass for fine-grained region control
- LowStar.RVector.fst: `vector` as a `regional` typeclass instance
- LowStar.Regional.Instances.fst: some useful typeclass instances

### CODE REVIEW

**Comment by joonwonc** [comment]
@msprotz and @tahina-pro : I would appreciate if you can take a look at the code and give some feedback!

File: ulib/LowStar.Regional.fst [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    // A representation value for the initial value of type `a`
+    irepr: Ghost.erased repr ->
+
+    // A property that should hold for all initial values of type `a`.
+    r_init_p: (a -> GTot Type0) ->
+
+    // An initialization operation. We might have a multiple ways of
+    // initializing a given target type `a`; then multiple typeclass instances
+    // should be defined, and each of them can be used properly.
+    r_init: (r:erid ->
```
Comment (msprotz) [inline ulib/LowStar.Regional.fst]: Hello Joonwon, Tahina and I are starting a quick review of this... sorry for the delay.
```suggestion
    r_alloc: (r:erid ->
```

Comment (msprotz) [reply ulib/LowStar.Regional.fst]: The rationale being that we generally use `init` for things that have been allocated (e.g. un-initialized buffers) already but for which we haven't written the initial state.

File: ulib/LowStar.Vector.fst [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+
+val as_seq: 
+  HS.mem -> #a:Type -> vec:vector a -> 
+  GTot (s:S.seq a{S.length s = U32.v (Vec?.sz vec)})
+let as_seq h #a vec =
+  B.as_seq h (B.gsub (Vec?.vs vec) 0ul (Vec?.sz vec))
+
+/// Capacity
+
+unfold val size_of: #a:Type -> vec:vector a -> Tot uint32_t
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: Why all the `unfold`s? I would suggest `inline_for_extraction` if you're hoping to see these helpers inlined at call-site

File: ulib/LowStar.Vector.fst [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (B.live h0 buf /\
+		  modifies (B.loc_buffer (B.gsub buf i len)) h0 h1))
+	(ensures (B.get h0 buf (U32.v k) == B.get h1 buf (U32.v k)))
+let get_preserved_buffer #a buf i len k h0 h1 =
+  get_as_seq_index h0 buf k; get_as_seq_index h1 buf k
+
+/// Abstract vector type
+
+noeq type vector_str a =
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: You might want to add `inline_for_extraction` on the type too, if you're hoping to see the projectors for the type inlined, too.

File: ulib/LowStar.Vector.fst [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+	    (Map.domain (MHS.get_hmap h1))
+
+val modifies_as_seq:
+  #a:Type -> vec:vector a -> dloc:loc ->
+  h0:HS.mem -> h1:HS.mem ->
+  Lemma (requires (live h0 vec /\ 
+		  loc_disjoint (loc_vector vec) dloc /\
+		  modifies dloc h0 h1))
+	(ensures (as_seq h0 vec == as_seq h1 vec))
+	[SMTPat (live h0 vec);
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: This pattern most likely won't work, owing to `live` defined as `unfold`

Comment (msprotz) [reply ulib/LowStar.Vector.fst]: the `unfold`s above are likely causing a bunch of issues, since `loc_vector` is also defined as unfold; Tahina and I recommend removing the `unfold`s above, and leaving `inline_for_extraction` for the small helpers you hope to see disappear when extracting; if there was a deeper reason for using `unfold`, let us know!

(the general rule of thumb is that a definition marked as unfold is inlined before going to SMT, meaning that it will not be able to be matched against)

Comment (msprotz) [reply ulib/LowStar.Vector.fst]: Tahina says it's a good practice to write `live h1 vec` in your post-condition because it's almost always going to be useful and will save z3 some search effort.

File: ulib/LowStar.Vector.fst [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+		 S.slice (as_seq h1 vec) (U32.v i) (U32.v j)))
+	[SMTPat (live h0 vec); 
+	SMTPat (loc_disjoint (loc_vector_within vec i j) dloc);
+	SMTPat (modifies dloc h0 h1)]
+let modifies_as_seq_within #a vec i j dloc h0 h1 =
+  B.modifies_buffer_elim (B.gsub (Vec?.vs vec) i (j - i)) dloc h0 h1
+
+/// Construction
+
+val create_empty:
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: this would be a good candidate for `inline_for_extraction`

File: ulib/LowStar.Vector.fst [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+	   live h1 vec /\ freeable vec /\
+	   modifies loc_none h0 h1 /\
+	   Set.equal (Map.domain (MHS.get_hmap h0))
+                     (Map.domain (MHS.get_hmap h1)) /\
+	   size_of vec = len /\
+	   S.equal (as_seq h1 vec) (S.create (U32.v len) v)))
+let create #a len v =
+  create_rid len v HH.root
+
+val create_reserve:
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: for this function, and possibly others around here, I would recommend writing a brief documentation that states in a single sentence what this function does compared to the other `create*` functions

File: ulib/LowStar.Vector.fst [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+	   frameOf vec = rid /\ live h1 vec /\ freeable vec /\
+	   modifies loc_none h0 h1 /\
+	   Set.equal (Map.domain (MHS.get_hmap h0))
+                     (Map.domain (MHS.get_hmap h1)) /\
+	   size_of vec = 0ul /\
+	   S.equal (as_seq h1 vec) S.empty))
+let create_reserve #a len ia rid =
+  Vec 0ul len (B.malloc rid ia len)
+
+val create_by_buffer:
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: Tahina recommends naming these `create*` based on the fact that `alloc`/`free` oftentimes go hand-in-hand

Comment (msprotz) [reply ulib/LowStar.Vector.fst]: (just like in the `Buffer` library)

File: ulib/LowStar.Vector.fst [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+       let nvs = B.malloc (B.frameOf vs) v ncap in
+       B.blit vs 0ul nvs 0ul sz;
+       B.upd nvs sz v;
+       B.free vs;
+       Vec (sz + 1ul) ncap nvs)
+  else
+    (B.upd vs sz v;
+    Vec (sz + 1ul) cap vs)
+
+val flush:
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: this one also could use a brief description

File: ulib/LowStar.Vector.fst [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+  f:(b -> a -> Tot b) -> ib:b ->
+  HST.ST b
+    (requires (fun h0 -> B.live h0 buf))
+    (ensures (fun h0 v h1 -> 
+      h0 == h1 /\
+      v == fold_left_seq (B.as_seq h0 buf) f ib))
+    (decreases (B.length buf))
+let rec fold_left_buffer #a #b len buf f ib =
+  if len = 0ul then ib
+  else (fold_left_buffer (len - 1ul) (B.sub buf 1ul (len - 1ul))
```
Comment (msprotz) [inline ulib/LowStar.Vector.fst]: This one could conceivably use the `C.Loops` module for a more efficient iteration, but that can come in a later pull request

Comment (msprotz) [reply ulib/LowStar.Vector.fst]: (especially since we need to move C.Loops to LowStar.Loops first)

File: ulib/LowStar.RVector.fst [hunk 10]
```diff
@@ -0,0 +1,1060 @@
...
+      MHS.get_hmap h1 == Map.upd (MHS.get_hmap h0) r1 Monotonic.Heap.emp /\
+      HH.color r1 = HH.color r0 /\
+      HyperStack.ST.is_eternal_region r1 /\
+      modifies loc_none h0 h1)
+let new_region_ r0 =
+  let hh0 = HST.get () in
+  let r1 = HST.new_region r0 in
+  let hh1 = HST.get () in
+  B.modifies_none_modifies hh0 hh1;
+  r1
```
Comment (msprotz) [inline ulib/LowStar.RVector.fst]: Tahina thinks that `new_region_modifies` from `LowStar.Monotonic.Buffer` should trigger. Do you know why you've had to define a separate helper?

**Comment by joonwonc** [comment]
Thanks a lot for the review!
I actually needed some changes in the `regional` typeclass and `rvector` in the meantime.
Let me try to apply both the changes and the review, and to commit a revision as soon as possible.

**Comment by msprotz** [comment]
The build is failing because of:
```
Error: Unbound type constructor FStar_Set.equal
```
I seem to recall this issue because you hit it earlier and @aseemr gave you a workaround. Does that ring a bell?

@tahina-pro can you do a pass on the updated changes!

Thanks for the fast turnaround time @joonwonc

**Comment by msprotz** [comment]
I'll let @aseemr tell us whether the workaround is acceptable for now or not -- I'd say yes.

**Comment by aseemr** [comment]
Re. `FStar_Set.ml:equal`, I suggested @joonwonc to add it, but with a comment that F* should not extract it. Once CMI related extraction changes are stable, we can revisit this.

**Comment by joonwonc** [comment]
I pushed the change to `FStar_Set.ml` with the comment.

**Comment by msprotz** [comment]
Perfect. Waiting for CI then I'll merge.

**Comment by msprotz** [comment]
@nikswamy  said he'd take care of the final merge... deferring to him. Thanks!

File: ulib/LowStar.Regional.fst [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+open FStar.All
+open FStar.Integers
+open FStar.Classical
+open LowStar.Modifies
+
+module HH = FStar.Monotonic.HyperHeap
+module HS = FStar.HyperStack
+module HST = FStar.HyperStack.ST
+module MHS = FStar.Monotonic.HyperStack
```
Comment (nikswamy) [inline ulib/LowStar.Regional.fst]: Why distinguish `MHS` from `HS`? The latter includes the former

Comment (joonwonc) [reply ulib/LowStar.Regional.fst]: Yes, no specific reason for this one; do you want me to fix it through all related files?

Comment (joonwonc) [reply ulib/LowStar.Regional.fst]: Fixed in the below commit!

**Comment by joonwonc** [comment]
I pushed one more fix in `LowStar.RVector` that caused memory leak.

**Comment by joonwonc** [comment]
Hello @nikswamy :

Thanks for the review, I really appreciate it.
Here are my answers for your questions:

> I don't understand how the representation chosen in LowStar.Vector.vector_str allows you to implement that This looks to me just like a thin wrapper around buffer allowing you to grow a buffer of a larger initial capacity than needed, allowing it to grow (up to the initial capactity) into the unused suffix.

Some operations of `vector` deal with reallocation. For instance, when `insert` is called, it tries to resize the capacity (by reallocating a buffer) when it's filled. This mechanism is abstracted so users don't need to worry about resizing while calling any operations.

> Can we have an interface for LowStar.Vector and LowStar.RVector?

I tried this, but found that a lot of predicate *definitions* are dependent to each other so they should be revealed in interface. If you think it's ok then I can work on it.

> Why the `dummy` element in the regional type class?

Sometimes we want to allocate (initialize) a data where an initial element is stateless (e.g., initializing a vector of buffers, filled with `null`). In order to have a general typeclass instance (e.g., `regional a -> regional (vector a)`) I had to have a general notion of `dummy` in `regional`.

> Why the separate `r_alloc_p: (a -> GTot Type0)`: Can't an instance just include whatever this predicate is meant to capture in the main `r_inv` invariant?

`r_alloc_p` is not an invariant, but a property that should hold for initial data. For instance, for `vector`s, `r_alloc_p` could say the length of the vector is zero. In my opinion, this property cannot be captured by `r_inv`.

> FStar.HyperStack.ST.erid is already defined: no need to duplicate it in Regional

Fixed.

> Should `region_of` in regional return an `erid`?

Currently it returns `rid` (`region_of: (a -> GTot HH.rid)`). I think `rid` is enough since it's mostly used to check the disjointness of `rid`s.

> Let's not open FStar.All, FStar.Classical etc. when they are not actually used

Fixed.

> Independently, a procedural issue: I'd like each file to be begin with the following copyright (especially since the work was done while at MSR) and license.

Fixed.

Let me know if you have more questions or suggestions!

File: ulib/LowStar.Vector.fst [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+      HST.is_eternal_region (frameOf vec)))
+    (ensures (fun h0 nvec h1 ->
+      frameOf vec = frameOf nvec /\
+      hmap_dom_eq h0 h1 /\
+      live h1 nvec /\ freeable nvec /\
+      modifies (loc_union (loc_addr_of_vector vec) 
+      			  (loc_vector nvec)) h0 h1 /\
+      size_of nvec = size_of vec + 1ul /\
+      get h1 nvec (size_of vec) == v /\
+      S.equal (as_seq h1 nvec) (S.snoc (as_seq h0 vec) v)))
```
Comment (nikswamy) [inline ulib/LowStar.Vector.fst]: @joonwonc Sorry for the slow reviews on this ...

I see now that `insert` resizes the vector, but it does by reallocating and passing back a resized vector, rather than mutating the original.

This is fine, but it is not the design goal described to me by @ad-l @protz and @fournet 

Is this expected by all?

Comment (joonwonc) [reply ulib/LowStar.Vector.fst]: I got your point, two things to mention here:

1. If you meant to maintain the original vector **struct** but to reallocate a **buffer**, this is possible by passing a vector struct **pointer** to the function and changing field values appropriately. As I remember, at the time I built this library we didn't have a way to use a field access with a struct pointer (`->`), so it currently returns a new vector struct. But as far as I know we can use `->` now so I could change the interface.

2. If you meant to maintain the original **buffer** pointer, we need a way to change the length of an allocated buffer. As far as I know we don't have such mechanism in C. So we just need to allocate a new buffer with a new capacity and to copy the content. This mechanism is also used by C++ std::vector, and for example, because of that, whenever we add en element to a vector, all iterators for the vector are invalidated (https://en.cppreference.com/w/cpp/container/vector/push_back).

Comment (msprotz) [reply ulib/LowStar.Vector.fst]: Passing a pointer to a struct then relying on KreMLin's pretty-printing to compile:

```
p.(0ul) <- { p.(0ul) with underlying_buffer with the_newly_allocated_buffer }
```

as:
```
p->underlying_buffer = the_newly_allocated_buffer
```
is always feasible on top of the current API, I believe (possibly with suitable inlining). However, this would be quite a burden to clients of this API, having to manage the outer pointer.

Since the struct is (I believe) three words in C, I would assume that making this function a struct-returning one, rather than an in-place mutating one, won't have too much of a performance impact.

I remember having similar discussions with @tahina-pro and also advocating for the struct-returning style, in order to limit the burden on clients.

So, in short, the current style is fine for me, and we can always define the more C-idiomatic style on top of it, should we want to.

Comment (nikswamy) [reply ulib/LowStar.Vector.fst]: I don't see why the client should have to explicitly manipulate a pointer to a struct with reliance on KreMLin for better code etc.

I was expecting the library to internally represent a vector as a pointer to a buffer, its current size and max capacity ... with buffer like operations (index, update) etc. lifted to this representation, and insert possibly resizing the buffer in case we're out of capacity. This should be completely transparent to clients.

**Comment by aseemr** [comment]
Hi @joonwonc. First off, thanks for the library and your patience with the review!

My main comment is about the use of `FStar.Monotonic.HyperHeap`. It is an internal module that clients should not use directly. `FStar.HyperStack` includes it, and so, all its symbols should be available under the `HS.` namespace. We left it there for backward compatibility but F* could give a warning if the clients open it directly. A better documentation could also help, currently there is just a comment in `FStar.Monotonic.HyperHeap`.

Could you make this change in the meantime.

**Comment by aseemr** [comment]
Could we also want a monotonic vector in the future? If so, this library should use `LowStar.Monotonic.Buffer` and should have a preorder.

We should merge it as is, but something to keep in mind. Also, changing it in the future would be easy enough.

**Comment by aseemr** [comment]
Oh another minor nit: the change in `FStar_Set.ml` is already in master (since @R1kM needed it for his development).

**Comment by joonwonc** [comment]
Hello @aseemr:

- I removed all use of `FStar.Monotonic.HyperHeap`.
- I'm not sure whether we will need monotonic vector for now, but as you said we can add the feature after merging, whenever needed.
- About `FStar_Set.ml`: I would prefer to resolve this possible conflict when merging, by choosing the file in `master`.

Thanks a lot for the review! Let me know if you have any more suggestions or questions.

**Comment by nikswamy** [comment]
I just merged this to master ... thanks Joonwon! Especially for your patience with the long discussion here.

The merge was overdue and other projects have been waiting for this code. But, there are still some aspects of it that can be improved.

You'll notice that I added a high-level description of the module to LowStar.Regional and LowStar.Vector.

It would be great to have comments of a similar flavor for LowStar.RVector and LowStar.Regional.Instances.

Another question is why the `copyable` class is in LowStar.RVector: doesn't it deserve its own file, like Regional? Also, can we have a more descriptive name instead of `RVector`? e.g., LowStar.Regional.Vector

Finally, do we really need a LowStar.Regional.Instances file? That's not particularly modular, e.g., as we add more instances. Is it feasible to give the instance for vector in LowStar.RVector/LowStar.Regional.Vector.

We could also have a LowStar.Regional.Buffer for the buffer instance.

**Comment by nikswamy** [comment]
As a bonus, it would be great to have another type, e.g., `LowStar.Resizable.Vector`, implemented as pointer to a LowStar.Vector, where resizing can be done imperatively rather than via struct passing (as discussed above and on slack).

**Comment by joonwonc** [comment]
@nikswamy : thanks for merging this pull request! Now I can work on using this library directly in my Merkle tree implementation in `hacl-star`.

I would be happy to improve the library with additional pull requests, so yes, I'd appreciate if you can make the pull request.
